### PR TITLE
telegram bot balance check

### PR DESCRIPTION
**Pull request by miiu96**
- Telegram bot to check balance of an address

### CODE REVIEW

File: tgbot/cmd/bot/config.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[config]
+    [config.general]
+        gatewayURL = "https://gateway.elrond.com"
+        address = "erd16x7le8dpkjsafgwjx0e5kw94evsqw039rwp42m2j9eesd88x8zzs75tzry"
+        balanceThreshold = "100000000000000000000000" # 100k EGLD
+        checkIntervalInMin = 30
+        notificationStep = 8 # num times to skip notifying if the balance check issue persists
+    [config.telegram]
+        groupID = ""
+        apiKey = ""
```
Comment (bogdan-rosianu) [inline tgbot/cmd/bot/config.toml:10]: empty line + replace address + gatewayURL

Comment (miiu96) [reply tgbot/cmd/bot/config.toml]: done.

**Review by bogdan-rosianu — APPROVED** [review]

File: tgbot/cmd/bot/main.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"io/ioutil"
+	"os"
+
+	logger "github.com/ElrondNetwork/elrond-go-logger"
```
Comment (bogdan-rosianu) [inline tgbot/cmd/bot/main.go:7]: rebrand the imports

File: tgbot/process/notifier.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	notificationStep   int
+
+	telegramBotKey  string
+	telegramGroupID string
+
+	notified bool
+	counter  int
+}
+
+func NewBalanceNotifier(cfg *config.GeneralConfig) (*notifier, error) {
```
Comment (bogdan-rosianu) [inline tgbot/process/notifier.go:31]: add comment

File: tgbot/process/notifier.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		balanceThreshold:   balanceThreshold,
+		checkIntervalInMin: cfg.BotConfig.General.CheckIntervalInMin,
+		notificationStep:   cfg.BotConfig.General.NotificationStep,
+
+		telegramBotKey:  cfg.BotConfig.Telegram.ApiKey,
+		telegramGroupID: cfg.BotConfig.Telegram.GroupID,
+	}, nil
+}
+
+func (n *notifier) StartNotifier() {
```
Comment (bogdan-rosianu) [inline tgbot/process/notifier.go:49]: add comment

File: tgbot/process/balance.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"errors"
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/blockchain"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/data"
+)
+
+func GetAddressBalance(gatewayURL string, address string) (*big.Int, error) {
```
Comment (bogdan-rosianu) [inline tgbot/process/balance.go:11]: add comment

File: tgbot/config/config.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package config
+
+type GeneralConfig struct {
```
Comment (bogdan-rosianu) [inline tgbot/config/config.go:3]: add comment

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: tgbot/cmd/bot/main.go (hunk 2 above)
Comment (miiu96) [reply tgbot/cmd/bot/main.go]: done

File: tgbot/process/notifier.go (hunk 3 above)
Comment (miiu96) [reply tgbot/process/notifier.go]: added

File: tgbot/process/notifier.go (hunk 4 above)
Comment (miiu96) [reply tgbot/process/notifier.go:60]: added

File: tgbot/process/balance.go (hunk 5 above)
Comment (miiu96) [reply tgbot/process/balance.go:14]: added

File: tgbot/go.mod [hunk 7]
```diff
@@ -0,0 +1,16 @@
+module github.com/ElrondNetwork/elrond-tools-go/tgbot
+
+go 1.15
```
[the remark this answers is not in the document]
Comment (miiu96) [reply tgbot/go.mod:3]: done

File: tgbot/process/notifier.go [hunk 8]
```diff
@@ -26,33 +32,47 @@ type notifier struct {
...
+		label:              cfg.General.Label,
 		balanceThreshold:   balanceThreshold,
-		checkIntervalInMin: cfg.BotConfig.General.CheckIntervalInMin,
-		notificationStep:   cfg.BotConfig.General.NotificationStep,
+		checkIntervalInMin: cfg.General.CheckIntervalInMin,
+		notificationStep:   cfg.General.NotificationStep,
 
-		telegramBotKey:  cfg.BotConfig.Telegram.ApiKey,
-		telegramGroupID: cfg.BotConfig.Telegram.GroupID,
+		telegramBotKey:  cfg.Telegram.ApiKey,
```
Comment (bogdan-rosianu) [inline tgbot/process/notifier.go:55]: remove empty line (L54)

File: tgbot/process/notifier.go [hunk 9]
```diff
@@ -86,14 +106,33 @@ func (n *notifier) checkBalanceAndNotifyIfNeeded() {
...
+	if res.StatusCode >= 400 {
+		b, _ := httputil.DumpResponse(res, true)
+		log.Warn(string(b))
+	}
+}
 
-	_, _ = http.Get(urlreq)
+// Close will close the notifier
+func (n *notifier) Close() error {
+	defer n.safeCloser.Close()
```
Comment (bogdan-rosianu) [inline tgbot/process/notifier.go:136]: could remove `defer`

File: tgbot/process/notifier.go (hunk 8 above)
Comment (miiu96) [reply tgbot/process/notifier.go:55]: done

File: tgbot/process/notifier.go (hunk 9 above)
Comment (miiu96) [reply tgbot/process/notifier.go]: done

**Review by bogdan-rosianu — APPROVED** [review]

**Review by schimih — APPROVED** [review]

**Pull request merged**